README.md
# Elixir Bme680

[![Build Status](https://travis-ci.org/lucaong/elixir_bme680.svg?branch=master)](https://travis-ci.org/lucaong/elixir_bme680) [![Hex Version](https://img.shields.io/hexpm/v/elixir_bme680.svg)](https://hex.pm/packages/elixir_bme680) [![docs](https://img.shields.io/badge/docs-hexpm-blue.svg)](https://hexdocs.pm/elixir_bme680/)

An Elixir library to interface with the BME680 and BME280 environmental sensors. The BME680
provides measurements of temperature, pressure, humidity, and gas resistance
(which is a proxy of indoor air quality). The ME680 is a lower cost device that only
provides measurements of temperature, pressure, humidity.

## Installation

The package can be installed
by adding `elixir_bme680` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:elixir_bme680, "~> 0.1.4"}
  ]
end
```

The Linux I2C driver needs to be installed for this library to work (e.g.
`libi2c-dev` on Debian). If using [Nerves](https://nerves-project.org), the
driver should already be installed by default.

## Usage with Bme680

```elixir
{:ok, pid} = Bme680.start_link()

measurement = Bme680.measure(pid)

# Measurement is like:
#
#   %Bme680.Measurement{
#     temperature: 21.74,
#     pressure: 1090.52,
#     humidity: 45.32,
#     gas_resistance: 10235
#   }
#
# Where temperature is in degrees Celsius, pressure in hPa, humidity in %
# relative humidity, and gas_resistance in Ohm
```

## Usage with Bme280

```elixir
{:ok, pid} = Bme280.start_link()

measurement = Bme280.measure(pid)

# Measurement is like:
#
#   %Bme280.Measurement{
#     temperature: 21.74,
#     pressure: 30.52,
#     humidity: 45.32
#   }
#
# Where temperature is in degrees Celsius, pressure in inHg, humidity in %
# relative humidity
```

For more information, read the [API documentation](https://hexdocs.pm/elixir_bme680).

## Sensor compatibility BME680

The default setting has been tested on the [Pimoroni
BME680](https://shop.pimoroni.com/products/bme680-breakout). The [Adafruit
BME680](https://www.adafruit.com/product/3660) requires using a different i2c
address. For the Adafruit, pass in the `i2c_address` option with value `0x77` as
follows:

```elixir
Bme680.start_link(i2c_address: 0x77)
```

## Sensor compatibility BME280

The default setting has been tested on the [HiLetgo
BME280](https://www.amazon.com/gp/product/B01N47LZ4P/).


## Note on gas resistance sensor warm up on the BME680

Note that, due to the nature of the BME680 gas resistance sensor, the gas
resistance measurement needs a warm-up in order to give stable measurements. One
possible strategy is to perform continuous meaurements in a loop until the value
stabilizes. That might take from a few seconds to several minutes (or more when
the sensor is brand new).

## Acknowledgements

This project contains low-level code from the [BME680 driver by
Bosch](https://github.com/BoschSensortec/BME680_driver) and the
[BME280 driver by Bosch](https://github.com/BoschSensortec/BME280_driver)


CHANGELOG.md
# Changelog

## v0.2.1

  - Update `elixir_make` development dependency, fixing an issue sometimes
    preventing the `priv` folder to be copied over to the device

## v0.2.0

  - Add support for BME280 (thanks to [@baldmountain](https://github.com/baldmountain))

## v0.1.4

  - Add configurable temperature compensation (`temperature_offset` option)

## v0.1.3

First stable version


lib/elixir_bme680.ex
defmodule Bme680 do
  @moduledoc """
  `elixir_bme680` provides a high level abstraction to interface with the
  BME680 environmental sensor on Linux platforms.
  """
  use Bitwise
  use GenServer

  defmodule State do
    @moduledoc false
    defstruct port: nil, subscribers: [], async_subscribers: [], measuring: false
  end

  defmodule Measurement do
    @moduledoc false
    defstruct temperature: nil, pressure: nil, humidity: nil, gas_resistance: nil
  end

  @doc """
  Starts and links the `Bme680` GenServer.

  Options:
    - `i2c_device_number` is the number of the i2c device, e.g. 1 for `/dev/i2c-1`
    - `i2c_address` i2c address of the sensor. It can be only `0x76` or `0x77`
    - `temperature_offset` is an offset, in degrees Celsius, that will be
      subtracted to temperature measurements in order to compensate for the internal
      heating of the device. It's typically around 4 or 5 degrees, and also
      affects relative humidity calculations
  """
  @spec start_link(
          [
            i2c_device_number: integer,
            i2c_address: 0x76 | 0x77,
            temperature_offset: non_neg_integer
          ],
          [term]
        ) :: GenServer.on_start()
  def start_link(bme_opts \\ [], opts \\ []) do
    i2c_device_number = Keyword.get(bme_opts, :i2c_device_number, 1)
    i2c_address = Keyword.get(bme_opts, :i2c_address, 0x76)
    temperature_offset = Keyword.get(bme_opts, :temperature_offset, 0)

    if Enum.member?([0x76, 0x77], i2c_address) do
      arg = [i2c_device_number, i2c_address, temperature_offset]
      GenServer.start_link(__MODULE__, arg, opts)
    else
      {:error, "invalid i2c address #{i2c_address}. Valid values are 0x76 and 0x77"}
    end
  end

  @doc """
  Perform a measurement on the BME680 sensor and synchronously return it

  Measurements are structs like:
  ```
  %Bme680.Measurement{
    temperature: 21.74,
    pressure: 1090.52,
    humidity: 45.32,
    gas_resistance: 10235
  }
  ```
  """
  @spec measure(GenServer.server()) :: %Measurement{
          temperature: float,
          pressure: float,
          humidity: float,
          gas_resistance: integer | nil
        }
  def measure(pid) do
    GenServer.call(pid, :measure)
  end

  @doc """
  Perform a measurement on the BME680 sensor and asynchronously send the result
  as a message to the pid in `send_to`
  """
  @spec measure_async(GenServer.server(), pid) :: :ok
  def measure_async(pid, send_to) do
    GenServer.cast(pid, {:measure_async, send_to})
  end

  @doc """
  Gracefully stops the `Bme680` GenServer.
  """
  @spec stop(GenServer.server()) :: :ok
  def stop(pid) do
    GenServer.cast(pid, :stop)
  end

  # GenServer callbacks

  def init([i2c_device_number, i2c_address, temperature_offset]) do
    executable_dir =
      Application.get_env(:elixir_bme680, :executable_dir, :code.priv_dir(:elixir_bme680))

    port =
      Port.open({:spawn_executable, executable_dir ++ '/bme680'}, [
        {:args, ["#{i2c_device_number}", "#{i2c_address}", "#{temperature_offset}"]},
        {:line, 64},
        :use_stdio,
        :binary,
        :exit_status
      ])

    {:ok, %State{port: port, measuring: false, subscribers: []}}
  end

  def handle_call(
        :measure,
        from,
        state = %State{port: port, subscribers: subscribers, measuring: measuring}
      ) do
    unless measuring, do: Port.command(port, "measure\n")
    {:noreply, %State{state | measuring: true, subscribers: [from | subscribers]}}
  end

  def handle_cast(
        {:measure_async, pid},
        state = %State{port: port, async_subscribers: subs, measuring: measuring}
      ) do
    unless measuring, do: Port.command(port, "measure\n")
    {:noreply, %State{state | measuring: true, async_subscribers: [pid | subs]}}
  end

  def handle_cast(:stop, state) do
    {:stop, :normal, state}
  end

  def handle_info({p, {:data, {:eol, line}}}, %State{
        port: p,
        subscribers: subs,
        async_subscribers: async_subs
      }) do
    measurement = decode_measurement(line)
    for pid <- subs, do: GenServer.reply(pid, measurement)
    for pid <- async_subs, do: send(pid, measurement)
    {:noreply, %State{port: p, measuring: false, subscribers: [], async_subscribers: []}}
  end

  def handle_info({port, {:exit_status, exit_status}}, state = %State{port: port}) do
    {:stop, exit_status, state}
  end

  # Private helper functions

  defp decode_measurement(line) do
    case line |> String.trim() |> String.split(",", trim: true) do
      ["T:" <> t, "P:" <> p, "H:" <> h, "G:" <> g] ->
        %Measurement{
          temperature: String.to_float(t),
          pressure: String.to_float(p),
          humidity: String.to_float(h),
          gas_resistance: String.to_integer(g)
        }

      ["T:" <> t, "P:" <> p, "H:" <> h] ->
        %Measurement{
          temperature: String.to_float(t),
          pressure: String.to_float(p),
          humidity: String.to_float(h)
        }

      _ ->
        {:error, "Measurement failed"}
    end
  end
end
